File: intellij-plugin/src/test/kotlin/dev/skillbill/intellij/composition/GoalMutationWiringTest.kt
package dev.skillbill.intellij.composition

import dev.skillbill.intellij.domain.SkillBillStatusOutcome
import dev.skillbill.intellij.fakes.FakePreferenceCache
import dev.skillbill.intellij.fakes.FakeStatusRepository
import dev.skillbill.intellij.fakes.ScriptedProcessFactory
import dev.skillbill.intellij.infrastructure.cli.ProcessRunner
import java.nio.file.Files
import java.time.Instant
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import org.junit.Assert.assertEquals
import org.junit.Test

class GoalMutationWiringTest {
    @Test
    fun `the composed pause and stop repositories each issue their own verb`() {
        val projectRoot = Files.createTempDirectory("mutation-wiring")
        val executable = Files.createTempFile("skill-bill", "")
        executable.toFile().setExecutable(true)
        val pauseFactory = ScriptedProcessFactory(exitCode = 0)
        val stopFactory = ScriptedProcessFactory(exitCode = 0)

        val composed = SkillBillStatusCompositionRoot.createForTest(
            projectRoot = projectRoot,
            preferences = FakePreferenceCache(cliOverride = executable.toString()),
            statusRepository = FakeStatusRepository { SkillBillStatusOutcome.Idle(Instant.now(), "idle") },
            pauseProcessRunner = ProcessRunner(processFactory = pauseFactory),
            stopProcessRunner = ProcessRunner(processFactory = stopFactory),
        )

        try {
            runBlocking {
                withContext(Dispatchers.Default) {
                    composed.goalPauseRepository.requestMutation(projectRoot, "SKILL-238")
                    composed.goalStopRepository.requestMutation(projectRoot, "SKILL-238")
                }
            }
            assertEquals(
                listOf(executable.toString(), "goal", "pause", "SKILL-238"),
                pauseFactory.commands.single().take(4),
            )
            assertEquals(
                listOf(executable.toString(), "goal", "stop", "SKILL-238"),
                stopFactory.commands.single().take(4),
            )
        } finally {
            composed.dispose()
        }
    }
}

File: vscode-extension/src/test/GoalMutationWiring.test.ts
import assert from "node:assert/strict";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { describe, it } from "mocha";
import { createGoalMutationRepositories } from "../composition/StatusCompositionRoot";
import { ProcessRunner } from "../infrastructure/cli/ProcessRunner";
import { FakePreferenceCache } from "./fakes/FakePreferenceCache";
import { ScriptedProcessFactory } from "./fakes/ScriptedProcessFactory";

class OverriddenCliPreferenceCache extends FakePreferenceCache {
  constructor(private readonly executable: string) {
    super();
  }

  override getCliExecutableOverride(): string {
    return this.executable;
  }
}

describe("GoalMutationWiring", () => {
  it("the composed pause and stop repositories each issue their own verb", async () => {
    const projectRoot = fs.mkdtempSync(path.join(os.tmpdir(), "mutation-wiring-"));
    const executable = path.join(projectRoot, "skill-bill");
    fs.writeFileSync(executable, "");
    fs.chmodSync(executable, 0o755);
    const pauseFactory = new ScriptedProcessFactory(0);
    const stopFactory = new ScriptedProcessFactory(0);

    const mutations = createGoalMutationRepositories(
      new OverriddenCliPreferenceCache(executable),
      new ProcessRunner(pauseFactory),
      new ProcessRunner(stopFactory),
    );

    await mutations.pause.requestMutation(projectRoot, "SKILL-238");
    await mutations.stop.requestMutation(projectRoot, "SKILL-238");

    assert.deepEqual(pauseFactory.commands[0]?.slice(0, 4), [executable, "goal", "pause", "SKILL-238"]);
    assert.deepEqual(stopFactory.commands[0]?.slice(0, 4), [executable, "goal", "stop", "SKILL-238"]);
  });
});
